vpn/tls.go

```diff
@@ -28,9 +28,57 @@ var (
 	ErrBadKeypair = errors.New("bad keypair conf")
 	// ErrBadParrot is returned for errors during TLS parroting
 	ErrBadParrot = errors.New("cannot parrot")
+	// ErrCannotVerifyCertChain is returned for certificate chain validation errors.
+	ErrCannotVerifyCertChain = errors.New("cannot verify chain")
 )
 
-// initTLS returns a tls.Config matching the VPN options.
+// certVerifyOptionsNoCommonNameCheck is a x509.VerifyOptions initialized with
+// an empty string for the DNSName. This allows to skip CN verification.
+var certVerifyOptionsNoCommonNameCheck = x509.VerifyOptions{DNSName: ""}
+
+// certVerifyOptions is the option that the customVerify function will use.
+var certVerifyOptions = certVerifyOptionsNoCommonNameCheck
+
+// customVerify is a version of the verification routines that does not try to verify
+// the Common Name, since we don't know it a priori for a VPN gateway. Returns
+// an error if the verification fails.
+// From tls/common documentation: If normal verification is disabled by
+// setting InsecureSkipVerify, [...] then this callback will be considered but
+// the verifiedChains argument will always be nil.
+func customVerify(rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
+	roots := x509.NewCertPool()
+	var leaf *x509.Certificate
+
+	for i, rawCert := range rawCerts {
+		cert, _ := x509.ParseCertificate(rawCert)
+		if cert != nil {
+			if i == 0 {
+				leaf = cert
+			} else {
+				roots.AddCert(cert)
+			}
+		}
+	}
+
+	opts := certVerifyOptions
+	opts.Roots = roots
+
+	var err error
```

Comment:
I think you don't need to declare `err` separately to when you use it. I suggest to drop this line and use `:= `below.

Reply:
refactor cruft, fixed